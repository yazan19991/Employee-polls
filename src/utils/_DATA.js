let users = {
  sarah_edo: {
    id: 'sarah_edo',
    name: 'Sarah Drasner',
    username: 'sarah_edo',
    password: 'sarah_edo',
    avatarURL: 'https://tylermcginnis.com/would-you-rather/sarah.jpg',
  },
  tylermcginnis: {
    id: 'tylermcginnis',
    name: 'Tyler McGinnis',
    username: 'tylermcginnis',
    password: 'tylermcginnis',
    avatarURL: 'https://tylermcginnis.com/would-you-rather/tyler.jpg',
  },
  dan_abramov: {
    id: 'dan_abramov',
    name: 'Dan Abramov',
    username: 'dan_abramov',
    password: 'dan_abramov',
    avatarURL: 'https://tylermcginnis.com/would-you-rather/dan.jpg',
  },
};

let polls = {
  '8xf0y6ziyjabvozdd253nd': {
    id: '8xf0y6ziyjabvozdd253nd',
    firstQuestion: 'Build our new application with Javascript',
    secondQuestion: 'Build our new applicaton with Typescript',
    author: 'sarah_edo',
    timestamp: 1679452943029,
    answers: [
      { choose: 2, user: 'tylermcginnis' },
      { choose: 1, user: 'dan_abramov' },
    ],
  },
  '5c9qojr2d1738zlx09afby': {
    id: '5c9qojr2d1738zlx09afby',
    firstQuestion: 'Build our new application with Spring',
    secondQuestion: 'Build our new applicaton with Spring Boot',
    author: 'sarah_edo',
    timestamp: 1518122597860,
    answers: [
      { choose: 1, user: 'sarah_edo' },
      { choose: 2, user: 'tylermcginnis' },
    ],
  },
  hbsc73kzqi75rg7v1e0i6a: {
    id: 'hbsc73kzqi75rg7v1e0i6a',
    firstQuestion: 'Build our new service with Java',
    secondQuestion: 'Build our new service with Python',
    author: 'tylermcginnis',
    timestamp: 1518122597860,
    answers: [
      { choose: 1, user: 'sarah_edo' },
      { choose: 1, user: 'tylermcginnis' },
    ],
  },
  '2mb6re13q842wu8n106bhk': {
    id: '2mb6re13q842wu8n106bhk',
    firstQuestion: 'Javascript is the most popular programming language',
    secondQuestion: 'Python is the most popular programming language',
    author: 'dan_abramov',
    timestamp: 1518122597860,
    answers: [
      { choose: 1, user: 'dan_abramov' },
      { choose: 2, user: 'tylermcginnis' },
      { choose: 2, user: 'sarah_edo' },
    ],
  },
  nnvkjqoevs8t02lzcc0ky: {
    id: 'nnvkjqoevs8t02lzcc0ky',
    firstQuestion: 'Build our new UI with React',
    secondQuestion: 'Build our new UI with Angular',
    author: 'sarah_edo',
    timestamp: 1679485709364,
    answers: [
      { choose: 2, user: 'dan_abramov' },
      { choose: 2, user: 'tylermcginnis' },
    ],
  },
};

let authedUser = {};

function generateUID() {
  return (
    Math.random().toString(36).substring(2, 15) +
    Math.random().toString(36).substring(2, 15)
  );
}

export function _getUsers() {
  return new Promise((res, rej) => {
    setTimeout(() => res({ ...users }), 1000);
  });
}

export function _getPolls() {
  return new Promise((res, rej) => {
    setTimeout(() => res({ ...polls }), 1000);
  });
}

export function _getAuthedUser() {
  return new Promise((res, rej) => {
    setTimeout(() => res(authedUser), 1000);
  });
}

export function _login(user) {
  authedUser = { ...authedUser, user };
  return new Promise((res, rej) => {
    setTimeout(() => res(user), 0);
  });
}

export function _saveVote({ id, answer }) {
  return new Promise((res, rej) => {
    if (!id || !polls[id]) {
      rej('Can not find poll');
    }

    setTimeout(() => {
      polls = {
        ...polls,
        [id]: {
          ...polls[id],
          answers: polls[id].answers.concat(answer),
        },
      };

      res();
    }, 0);
  });
}

function formatPoll(firstQuestion, secondQuestion, author) {
  return {
    id: generateUID(),
    firstQuestion,
    secondQuestion,
    author,
    timestamp: Date.now(),
    answers: [],
  };
}

export function _savePoll(firstQuestion, secondQuestion, author) {
  return new Promise((res, rej) => {
    if (!firstQuestion || !secondQuestion || !author) {
      rej('The questions and author are required');
    }

    const poll = formatPoll(firstQuestion, secondQuestion, author);

    setTimeout(() => {
      polls = {
        ...polls,
        [poll.id]: poll,
      };

      res(poll);
    }, 500);
  });
}

export function _logout() {
  return new Promise((res, rej) => {
    setTimeout(() => {
      authedUser = {};
      res();
    }, 0);
  });
}
