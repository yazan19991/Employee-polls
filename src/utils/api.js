import {
  _getUsers,
  _getPolls,
  _getAuthedUser,
  _login,
  _saveVote,
  _savePoll,
  _logout,
} from './_DATA';

export function getUsers() {
  return _getUsers();
}

export function getPolls() {
  return _getPolls();
}

export function getAuthedUser() {
  return _getAuthedUser();
}

export function login(user) {
  return _login(user);
}

export function saveVote({ id, answer }) {
  return _saveVote({ id, answer });
}

export function savePoll(firstQuestion, secondQuestion, author) {
  return _savePoll(firstQuestion, secondQuestion, author);
}

export function logout() {
  return _logout();
}
