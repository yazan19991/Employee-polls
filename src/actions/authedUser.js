import { login, getAuthedUser, logout } from '../utils/api';

export const LOGIN = 'LOGIN';
export const RECEIVE_AUTHED_USER = 'RECEIVE_AUTHED_USER';
export const LOGOUT = 'LOGOUT';

function doLogin(authedUser) {
  return {
    type: LOGIN,
    authedUser,
  };
}

export function handleLogin(user) {
  return (dispatch) => {
    return login(user).then((authedUser) => dispatch(doLogin(authedUser)));
  };
}

function receiveAuthedUser(authedUser) {
  return {
    type: RECEIVE_AUTHED_USER,
    authedUser,
  };
}

export function handleReceiveAuthedUser() {
  return (dispatch) => {
    return getAuthedUser().then((authedUser) =>
      dispatch(receiveAuthedUser(authedUser))
    );
  };
}

function doLogout() {
  return {
    type: LOGOUT,
    authedUser: {},
  };
}

export function handleDoLogOut() {
  return (dispatch) => {
    return logout().then(() => dispatch(doLogout()));
  };
}
