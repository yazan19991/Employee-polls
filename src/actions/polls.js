import { getPolls, saveVote, savePoll } from '../utils/api';

export const RECEIVE_POLLS = 'RECEIVE_POLLS';
export const VOTE = 'VOTE';
export const CREATE_POLL = 'CREATE_POLL';

function receivePolls(polls) {
  return {
    type: RECEIVE_POLLS,
    polls,
  };
}

export function handleReceivePolls() {
  return (dispatch) => {
    return getPolls().then((polls) => dispatch(receivePolls(polls)));
  };
}

function vote(id, answer) {
  return {
    type: VOTE,
    id,
    answer,
  };
}

export function handleVote(id, answer) {
  return (dispatch) => {
    return saveVote({ id, answer }).then(() => dispatch(vote(id, answer)));
  };
}

function createPoll(poll) {
  return {
    type: CREATE_POLL,
    poll,
  };
}

export function handleCreatePoll(firstQuestion, secondQuestion, author) {
  return (dispatch) => {
    return savePoll(firstQuestion, secondQuestion, author).then((poll) =>
      dispatch(createPoll(poll))
    );
  };
}
