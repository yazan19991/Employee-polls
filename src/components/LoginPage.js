import { connect } from 'react-redux';
import { useState } from 'react';
import { useNavigate, useLocation } from 'react-router-dom';
import { handleLogin } from '../actions/authedUser';
import logo from '../images/login-background.png';

const LoginPage = (props) => {
  const navigate = useNavigate();
  const location = useLocation();

  if (props.authedUser && props.authedUser.id) {
    navigate('/home');
  }

  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [loginFail, setLoginFail] = useState(false);

  const handleOnChangeUsername = (e) => {
    const username = e.target.value;
    setUsername(username);
  };

  const handleOnChangePassword = (e) => {
    const password = e.target.value;
    setPassword(password);
  };

  const handleOnSubmit = (e) => {
    e.preventDefault();

    let authedUser = { ...props.users[username] };
    if (authedUser.password !== password) {
      authedUser = undefined;
      setLoginFail(true);
      return;
    }

    props.dispatch(handleLogin(authedUser));
    setLoginFail(false);
    setUsername('');
    setPassword('');
    navigate(location.state?.path || '/home');
  };

  return (
    <div className="container">
      <img src={logo} className="login-img" alt="survey-background" />
      <h2>Log In</h2>
      <form className="container" onSubmit={handleOnSubmit}>
        <h3>User</h3>
        <input
          data-testid="username"
          type="text"
          className="input"
          placeholder="username"
          value={username}
          onChange={handleOnChangeUsername}
        />
        <h3>Password</h3>
        <input
          data-testid="password"
          type="password"
          className="input"
          placeholder="password"
          value={password}
          onChange={handleOnChangePassword}
        />
        {loginFail && (
          <p data-testid="error-msg" className="error-message">
            Invalid username or password
          </p>
        )}
        <button
          data-testid="login-btn"
          type="submit"
          className="btn-submit"
          disabled={username === '' || password === ''}
        >
          Login
        </button>
      </form>
    </div>
  );
};

const mapStateToProps = ({ authedUser, users }) => ({
  authedUser,
  users,
});

export default connect(mapStateToProps)(LoginPage);
