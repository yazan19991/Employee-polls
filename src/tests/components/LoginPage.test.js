/* eslint-disable testing-library/prefer-screen-queries */
import { fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom';
import LoginPage from '../../components/LoginPage';
import { MemoryRouter } from 'react-router-dom';
import { createStore } from 'redux';
import reducer from '../../reducers';
import middleware from '../../middleware';
import { Provider } from 'react-redux';
import { getUsers } from '../../utils/api';
import { receiveUsers } from '../../actions/users';
import { act } from 'react-dom/test-utils';

const store = createStore(reducer, middleware);

describe('Testing Login component', () => {
  beforeAll(async () => {
    await getUsers().then((data) =>
      act(() => {
        store.dispatch(receiveUsers(data));
      })
    );
  });

  it('Will match snapshot', () => {
    const { component } = render(
      <Provider store={store}>
        <MemoryRouter>
          <LoginPage />
        </MemoryRouter>
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });

  it('Will display error message if username is wrong', () => {
    // eslint-disable-next-line testing-library/render-result-naming-convention
    const component = render(
      <Provider store={store}>
        <MemoryRouter>
          <LoginPage />
        </MemoryRouter>
      </Provider>
    );
    const usernameInput = component.getByTestId('username');
    fireEvent.change(usernameInput, { target: { value: 'hello' } });
    const passwordInput = component.getByTestId('password');
    fireEvent.change(passwordInput, { target: { value: 'sarah_edo' } });
    const loginBtn = component.getByTestId('login-btn');
    fireEvent.click(loginBtn);
    expect(component.getByTestId('error-msg')).toBeInTheDocument();
  });

  it('Will display error message if password is wrong', () => {
    // eslint-disable-next-line testing-library/render-result-naming-convention
    const component = render(
      <Provider store={store}>
        <MemoryRouter>
          <LoginPage />
        </MemoryRouter>
      </Provider>
    );
    const usernameInput = component.getByTestId('username');
    fireEvent.change(usernameInput, { target: { value: 'sarah_edo' } });
    const passwordInput = component.getByTestId('password');
    fireEvent.change(passwordInput, { target: { value: 'hello' } });
    const loginBtn = component.getByTestId('login-btn');
    fireEvent.click(loginBtn);
    expect(component.getByTestId('error-msg')).toBeInTheDocument();
  });

  it('Will display error message if username and password is match', () => {
    // eslint-disable-next-line testing-library/render-result-naming-convention
    const component = render(
      <Provider store={store}>
        <MemoryRouter>
          <LoginPage />
        </MemoryRouter>
      </Provider>
    );
    const usernameInput = component.getByTestId('username');
    fireEvent.change(usernameInput, { target: { value: 'sarah_edo' } });
    const passwordInput = component.getByTestId('password');
    fireEvent.change(passwordInput, { target: { value: 'sarah_edo' } });
    const loginBtn = component.getByTestId('login-btn');
    fireEvent.click(loginBtn);
    expect(component.queryByTestId('error-msg')).not.toBeInTheDocument();
  });
});
