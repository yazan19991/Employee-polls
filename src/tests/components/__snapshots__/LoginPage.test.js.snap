// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Login component Will match snapshot 1`] = `undefined`;
